### ISSUE TITLE
Cannot change keyboard layout while on wayland session.

### ISSUE THREAD

**Issue by Cristiandis** [linuxmint/wayland#14]
### Distribution

Linux Mint 22

### Package version

6.2.0

### Graphics hardware in use

Intel Corporation GeminiLake [UHD Graphics 600]

### Frequency

Always

### Bug description

Missing keyboard layout option while on wayland session.

### Steps to reproduce

Open settings and go to keyboard.

### Expected behavior

There should be a third button that open a menu where you can change the keyboard layout but its missing in wayland session.

**Comment by Leniwcowaty** [comment]
Can confirm, Linux Mint 21.3, no Keyboard Layout in Wayland and defaults to English, issue not present on X11

**Comment by stvoidit** [comment]
I will subscribe to this question.
So far, this problem is the only one for which I cannot fully use the Wayland session and test whenever possible.

**Comment by 1vanK** [comment]
> So far, this problem is the only one for which I cannot fully use the Wayland session and test whenever possible.

I also think this should be a higher priority.

**Comment by RossBrunton** [comment]
Dvorak user (btw) here. This is a blocker for me as well. I want to use Wayland (my GPU I think doesn't play nicely with X), but can't because of this.

**Comment by WhenGryphonsFly** [comment]
I regularly use the compose key (bound to scroll lock). I also have a custom XKB layout that classifies minus/underscore as alphabetic so that it is affected by caps lock. My understanding is that Wayland uses (or at least can use) XKB for keyboard layouts. Therefore, it is very annoying that even when an XKB layout (or compose key position) is manually set via dconf, the layout is not respected.

While I could struggle through the lack of underscore in caps lock, the lack of a compose key is a blocker for me. I believe I could also deal with the lack of ctrl+click (see issue 1), making this the _sole_ blocker for me switching to Wayland for full-time testing.

I see that in #16 (before it was closed as a duplicate) that there is no timeline for work on this issue. I would be willing to work on this issue myself, but unfortunately I have too many other obligations to devote time to learning Cinnamon's codebase. Therefore, all I can ask is that the priority of this issue be reevaluated.

**Comment by ultraph** [comment]
will wait 4 this issue resolution.

**Comment by Bundy01** [comment]
Hi,

I would add that my AZERTY keyboard is QWERTY on some apps (Firefox, settings... No problem with Brave-browser,KeepassXC...).

For my root password, I use an `@` that turns into `<ffffff>`. So I can't administer the PC with Wayland, even if I copy the password :(

Regards.

**Comment by Cristiandis** [comment]
Still not addressed in 6.2 with mint 22

**Comment by mruncreative** [comment]
This was the first thing I noticed negatively starting a Wayland session on Mint 22. Everything looked all right in Cinnamon until I started typing.

**Comment by ZeroOVi** [comment]
Yeah, i cant change keyboard layout too (spanish keyboard)

**Comment by testsieger73** [comment]
Can this be pushed as it makes Wayland useless for international users

**Comment by mtwebster** [comment]
Wayland support in Cinnamon is entirely experimental. You use it at your own risk. We *definitely* don't recommend it as a daily driver.

This issue tracker is a way for those who are interested in testing it can provide feedback. It is not necessarily going to inform any decisions we make regarding our development priorities.

**Comment by Weranders** [comment]
While I find this understandable, not having support for other keyboard layouts than the english one drastically limits the pool of early adopters and testers.

I just started using Wayland, fully aware that is is experimental and with helping in testing in the back of my head. Everything was fine, even better than x11 since I was finally able to configure my screen setup how I liked it which always failed before. 

Until I started typing something. I will have to revert now until this issue is fixed.

**Comment by testsieger73** [comment]
Push

**Comment by KAMI911** [comment]
Unfortunately this makes hard to use Wayland session for Hungarian users. It should be handled as a high priority or blocker bug. Thank you!

**Comment by ROBERT-MCDOWELL** [comment]
We all understand it's experimental, but how to experiment if at least we cannot experiment with a keyboard layout ????

**Comment by silverhadch** [comment]
This is probablly the highest priority bug in this experimental wayland session. As the international FOSS Community cant use and test it. ???????

**Comment by ROBERT-MCDOWELL** [comment]
because of that I cannot experiment wayland so I'm back to the slow Xorg...

**Comment by dj0abr** [comment]
I gave wayland a try and it worked much better than expected, unfortunately the keyboard layout is alway US, so I cannot continue testing.

**Comment by nrv9999** [comment]
I also noticed that the Shift and Ctrl buttons did not work in the game (cs2) in wayland session (X ok). Maybe it's related.

**Comment by drinfohouse** [comment]
Hi folks, testing wayland on Thinkpad T430. 
It's working quite well but the keyboard issue happened. 

Since i'm a brasilian user, I need to have the keyboard on PT-BR to at least test / do something. 
This is the first really bad issue that blocked me to continue using wayland since i can't type as necessary. 
I would love to keep using/testing wayland on cinnamon and provide support if this issue is solved.

**Comment by ROBERT-MCDOWELL** [comment]
@drinfohouse 
welcome to the rest of the world club...

**Comment by duneatnight** [comment]
I can't believe this is went into Mint 22

**Comment by KAMI911** [comment]
I hope they will fix for Mint 22.1.

**Comment by abda11ah** [comment]
stuck in X11 because i need an AZERTY keyboard layout.

**Comment by silverhadch** [comment]
> I can't believe this is went into Mint 22

They said because slow adoption and stability, however if i cant use your damn Software at all then it doesnt matter.

**Comment by Fantu** [comment]
Information about a connected thing: libgnomekbd used for X11 part is no longer maintained upstream and archived

**Comment by drinfohouse** [comment]
22.1 is expected to fix this issue? I'm so lost :(

**Comment by ROBERT-MCDOWELL** [comment]
we need to open an issue to the wayland repo

**Comment by drinfohouse** [comment]
you mean not this repo, rigth?

**Comment by ROBERT-MCDOWELL** [comment]
yes, everyone should open an issue on wayland repo as it's a wayland protocol issue, many linux distributions have this issue.... https://gitlab.freedesktop.org/wayland

**Comment by Fantu** [comment]
https://gitlab.freedesktop.org/wayland/wayland/-/issues/428

**Comment by mtwebster** [comment]
> yes, everyone should open an issue on wayland repo as it's a wayland protocol issue, many linux distributions have this issue.... https://gitlab.freedesktop.org/wayland

> https://gitlab.freedesktop.org/wayland/wayland/-/issues/428

### This is NOT why you can't currently change keyboard layouts in Cinnamon

**Comment by Fantu** [comment]
The link posted was only for one developer reply probably useful for who would say that it is a Wayland bug while unlike how you are perhaps accustomed to X11 the implementations with wayland the implementation is made mainly in the compositors

**Comment by mtwebster** [comment]
@Fantu no worries, I just don't want casual folks seeing this and confusing this specific Cinnamon issue with 'at large' issues/development in Wayland, especially those users who might tend to get a bit over-dramatic about things.

**Comment by silverhadch** [comment]
> yes, everyone should open an issue on wayland repo as it's a wayland protocol issue, many linux distributions have this issue.... https://gitlab.freedesktop.org/wayland

BS

**Comment by ROBERT-MCDOWELL** [comment]
SO if it's BS, explaine why after 2 years no update, no fix?, be smart rather than insult.

**Comment by DAMEK86** [comment]
Sounds like after almost over 1 year, there is no effort to fix this issue but within a view days you where able to point to Wayland?

Sad to hear. Because I like mint a lot. However, not being able to use my keyboard makes it useless for me and the current version lacks some newer kernels because of the newer gcc. :(
Hope you can rethink the issue since the rest of the world likes your software as well so far.

**Comment by ROBERT-MCDOWELL** [comment]
again this is not a linux distribution like mint as it's present on ALL cinnamon desktop mamagers, including fedora, etc...

**Comment by RossBrunton** [comment]
To distract from blame game going on here, what components actually need porting over for this to be supported?

Just for testing while it's experimental, I'd be fine with setting it in a config file or dconf config that is read once on startup.

**Comment by ROBERT-MCDOWELL** [comment]
yeah is anyone has any info of how to contribute to make this going faster?

**Comment by silverhadch** [comment]
> yeah is anyone has any info of how to contribute to make this going faster?

Well they forked a compositor which already has it so maybe we just look at the implementation of the forked Compositor.

**Comment by ROBERT-MCDOWELL** [comment]
is "they" linuxmint? or ?

**Comment by mtwebster** [comment]
I'm going to lock this, there's nothing useful to be added here, specific to Cinnamon and Wayland, and the noise is unwelcome.

Much of what's required for keyboard layout changes exists already in muffin, this just needs implementing in Cinnamon itself. I should have some progress made here for Cinnamon 6.6.

**Comment by mtwebster** [comment]
For anyone interested in testing a WIP for this:

https://github.com/linuxmint/cinnamon/pull/12758

Note the required additonal PR linked there also.

Normal keyboard layout switching should work now (not ibus/fcitx, etc... nor have I tested cinnamon's virtual keyboard with any of it - this is all stuff still in progress).

Use at your own risk. If you notice any issues feel free to mention on that PR, but I can't offer any real support for this currently.

**Comment by mtwebster** [comment]
All PRs but one (mintlocale) merged.
ref: https://github.com/linuxmint/cinnamon/pull/12758
